packages/editor/src/components/CodeEditor/ExpressionEditor.tsx

```diff
@@ -222,7 +228,14 @@ export const BaseExpressionEditor: React.FC<EditorProps> = ({
       cm.current?.off('change', changeHandler);
       cm.current?.off('blur', blurHandler);
     };
-  }, [defaultCode, onChange, onBlur, compact]);
+  }, [code, onChange, onBlur, compact]);
+  useEffect(() => {
+    if (cm.current) {
+      if (code !== cm.current.getValue()) {
+        cm.current.setValue(code);
```

Comment:
The reason to do that is the API's URL input and parameter inputs all use `ExpressionEditor`, and then the URL and parameter inputs would need to sync values to each other.

> I'm not sure whether this will destroy CodeMirror's context since it maintains a lot of complex state and context internally.

Maybe not? I check the codes of these implements, they just depend on the `change` event of CodeMirror which also would be triggered by calling `cm.current.setValue(code)`.  
Or maybe you worry about some developers would change `ExpressionEditor`'s `code` while the user is inputting and this would change the cursor location to break the CodeMirror's context? the developer who uses `ExpressionEditor` probably doesn't do this, because it also would break the user input action which is awful.

Reply:
This looks a little implicit because a react component with `[value, onChange]` props set looks like a controlled component.
The current implementation/usage looks like an uncontrolled component with reactive default value props.

Anyway, I think this is a problem of react's design patterns, which discourages users implement imperative API, which CodeMirror does(`cm.current.setValue`).

Two workarounds may help:
1. [Add an imperative API to the `ExpressionEditor` via `useImperativeHandle `](https://medium.com/condorlabs-engineering/exposing-functions-from-react-functional-components-1b77c32484bd).
2. Add a key to the `ExpressionEditor`, when we need a force value update to the editor update the key and let the editor remount. This follows react's patterns but may introduce performance overhead.

Reply:
OK. I change to expose `setCode` from `ExpressionEditor` and revert the `code` to `defaultCode` now.